### DISCUSSION TITLE
Hi I need help, CUDA out of memory. How to solve this issue?

### DISCUSSION THREAD

**Discussion by JenuelDev** [milvus-io/milvus#33626] [Q&A and General discussion]
This is the error coming out when I try to insert thousands of data to collection.
```
CUDA out of memory. Tried to allocate 20.00 MiB. GPU 0 has a total capacity of 15.73 GiB of which 19.12 MiB is free. Process 205777  
 has 4.18 GiB memory in use. Process 281977 has 1.34 GiB memory in use. Including non-PyTorch memory, this process has 10.19 GiB 
 memory in use. Of the allocated memory 8.52 GiB is allocated by PyTorch, and 662.33 MiB is reserved by PyTorch but unallocated. If  
 reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid  
 fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment- 
 variables)
```

**Comment by yhmo** [comment]
I guess the GPU memory is occupied by other processes.

**Reply by JenuelDev** [reply]
Its throwing error when I try to insert thousands of data to a collection.

**Reply by JenuelDev** [reply]
Their are no other processess

**Reply by JenuelDev** [reply]
their is only process but only uses 4gb
![image](https://github.com/milvus-io/milvus/assets/31676163/e05d22d4-e490-489c-9a75-f3d2e5f2dda5)

**Reply by yhmo** [reply]
Read the error message and you will see what the CUDA said:
**CUDA out of memory. Tried to allocate 20.00 MiB. GPU 0 has a total capacity of 15.73 GiB of which 19.12 MiB is free. 
Process 205777 has 4.18 GiB memory in use. 
Process 281977 has 1.34 GiB memory in use. 
Including non-PyTorch memory, this process has 10.19 GiB memory in use. 
Of the allocated memory 8.52 GiB is allocated by PyTorch, and 662.33 MiB is reserved by PyTorch but unallocated.**

CUDA doesn't lie.

**Reply by JenuelDev** [reply]
hmm,,, it seems that Its running no process.. but when I do some milvus getting data to milvus its using every GPU.

**Comment by xiaofan-luan** [comment]
GPU 0 has a total capacity of 15.73 GiB of which 19.12 MiB is free.

I guess this is clearly that all memory is all allocated out.
to be noticed, when use GPU index, the GPU did a lot of preallocations of memory.  I think it's simply easier to not sharing GPU memory between pytorch and milvus. 

Is there any special reason we want to use GPU?

**Reply by JenuelDev** [reply]
Because GPU is a lot fast I guess?,, its what our lead recommended to use the GPU.

**Reply by xiaofan-luan** [reply]
> Because GPU is a lot fast I guess?,, its what our lead recommended to use the GPU.

We don't recommend to use GPU for vector search until you have many knowledge. most likely cpu will be fast enough. Leave you GPU for model inference make more sense.
Unless you are running batch query or you have high QPS -> more than 1000QPS, use cpu will be good enought

**Reply by JenuelDev** [reply]
by default its using GPU,, how to use CPU?

**Reply by xiaofan-luan** [reply]
if you use cpu image, all the index is cpu based. as long as you don't use GPU related index listed in docs

**Reply by JenuelDev** [reply]
can you tell me more about it?

**Reply by xiaofan-luan** [reply]
not sure what do you want do you want to know. https://milvus.io/docs almost everything here is about CPU expect for "Run milvus with GPU"
